pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -61,172 +103,188 @@ type ecrTokenGetter struct {
+func stringKeyFunc(obj interface{}) (string, error) {
+	key := obj.(cacheEntry).host
+	return key, nil
 }
 
-// Enabled implements DockerConfigProvider.Enabled for the lazy provider.
-// Since we perform no checks/work of our own and actualProvider is only created
-// later at image pulling time (if ever), always return true.
-func (p *lazyEcrProvider) Enabled() bool {
-	return true
+func (p *ecrExpirationPolicy) IsExpired(entry *cache.TimestampedEntry) bool {
+	expiresAt := entry.Obj.(cacheEntry).expiresAt
+	return expiresAt.Before(time.Now()) //TODO clear before expiration
 }
 
-// LazyProvide implements DockerConfigProvider.LazyProvide. It will be called
-// by the client when attempting to pull an image and it will create the actual
-// provider only when we actually need it the first time.
-func (p *lazyEcrProvider) LazyProvide() *credentialprovider.DockerConfigEntry {
-	if p.actualProvider == nil {
-		klog.V(2).Infof("Creating ecrProvider for %s", p.region)
-		p.actualProvider = &credentialprovider.CachingDockerConfigProvider{
-			Provider: newEcrProvider(p.region, nil),
-			// Refresh credentials a little earlier than expiration time
-			Lifetime: 11*time.Hour + 55*time.Minute,
-		}
-		if !p.actualProvider.Enabled() {
-			return nil
-		}
+/*GET CREDENTIALS*/
+func (p *ecrProvider) getFromCache(parsed *parsedURL) (credentialprovider.DockerConfig, bool) {
+	klog.Infof("Checking cache for credentials for %v", parsed.host)
+	cfg := credentialprovider.DockerConfig{}
+	obj, exists, err := p.cache.GetByKey(parsed.host)
+	if err != nil {
+		klog.Infof("unable to get credentials from cache for %v %v", parsed.host, err)
+		return cfg, exists
 	}
-	entry := p.actualProvider.Provide()[p.regionURL]
-	return &entry
+	if exists {
+		entry := obj.(cacheEntry)
+		klog.Info("Credentials found in cache")
+		fmt.Println("in cache")
+		cfg[entry.host] = entry.credentials
+	} else {
+		fmt.Println("not in cache") //TODO
+	}
+	return cfg, exists
 }
 
-// Provide implements DockerConfigProvider.Provide, creating dummy credentials.
-// Client code will call Provider.LazyProvide() at image pulling time.
-func (p *lazyEcrProvider) Provide() credentialprovider.DockerConfig {
-	entry := credentialprovider.DockerConfigEntry{
-		Provider: p,
-	}
+func (p *ecrProvider) getFromECR(parsed *parsedURL) credentialprovider.DockerConfig {
+	klog.Infof("Getting credentials from ECR for %v", parsed.host)
 	cfg := credentialprovider.DockerConfig{}
-	cfg[p.regionURL] = entry
+	if p.getter == nil {
+		p.getter = newGetter(parsed.region)
+	}
+
+	params := &ecr.GetAuthorizationTokenInput{RegistryIds: []*string{aws.String(parsed.registryID)}}
+	output, err := p.getter.GetAuthorizationToken(params)
+	if err != nil {
+		klog.Errorf("while requesting ECR authorization token %v", err)
+		return cfg
+	}
+	if output == nil {
+		klog.Errorf("Got back no ECR token")
+		return cfg
+	}
+
+	data := output.AuthorizationData[0]
+	if data.ProxyEndpoint != nil &&
+		data.AuthorizationToken != nil {
```

Comment:
The check for `data.AuthorizationToken != nil` is the most important part of this. `data.ProxyEndpoint` being nil just means that you'll log a nil, right? If that's true then may as well remove the first condition.

Also this seems an odd case that ECR would not return a credential yet also not error. I think it would be worth an else block just log `Errorf` to make debugging easier later.

Reply:
I need to look into this -- it was from the original code.

Reply:
Should I also return if proxyendpoint is nil?  then it won't try doing the rest.

Reply:
Think I should just return if ProxyEndpoint is nil? And it'll log a nil endpoint and the error. 

```
if len(output.AuthorizationData) == 0 {
    klog.Errorf("Got back no ECR authorization data")
    return cfg
}
data := output.AuthorizationData[0]
if data.AuthorizationToken == nil {
    klog.Errorf("Got back no ECR authorization token")
    return cfg
}
if data.ProxyEndpoint == nil {
    klog.Errorf("Got back no ECR proxy endpoint")
    return cfg
}
decodedToken, err := base64.StdEncoding.DecodeString(aws.StringValue(data.AuthorizationToken))
...
```